File: SchoolParsers/Section.cpp
#include "Section.hpp"
#include "Instructor.hpp"
#include "Week.hpp"
#include "Semester.hpp"
#include "Location.hpp"
#include <string>

/*
 * ======================================================
 * Object Creation
 * ======================================================
 */

Section::Section(){
	this->_sectionName = "";
	this->_sectionType = "";
	this->_description = "";
	this->_crn = "";

	this->_instructor 	= NULL;
	this->_daysOfWeek 	= NULL;
	this->_dates 		= NULL;
	this->_bulding 		= NULL;
}

Section::~Section(){
	delete this->_instructor;
	delete this->_daysOfWeek;
	delete this->_dates;
	delete this->_bulding;
}

Section::Section(const Section& copy){
	this->_sectionName = copy._sectionName;
	this->_sectionType = copy._sectionType;
	this->_description = copy._description;
	this->_crn 		   = copy._crn;

	*(this->_instructor) = *(copy._instructor);
	*(this->_daysOfWeek) = *(copy._daysOfWeek);
	*(this->_dates) 	 = *(copy._dates);
	*(this->_bulding)	 = *(copy._bulding);
}

Section& Section::operator=(const Section& copy){
	this->_sectionName = copy._sectionName;
	this->_sectionType = copy._sectionType;
	this->_description = copy._description;
	this->_crn         = copy._crn;

	*(this->_instructor) = *(copy._instructor);
	*(this->_daysOfWeek) = *(copy._daysOfWeek);
	*(this->_dates) 	 = *(copy._dates);
	*(this->_bulding)	 = *(copy._bulding);

	return *this;

}

/*
 * ======================================================
 * Setters
 * ======================================================
 */


void Section::setSectionName(string sectionName){
	this->_sectionName = sectionName;
}

void Section::setSectionType(string sectionType){
	this->_sectionType = sectionType;
}

void Section::setDescription(string description){
	this->_description = description;
}

void Section::setInstructor(Instructor* instructor){
	this->_instructor = instructor;
}

void Section::setDaysOfWeek(Week* daysOfWeek){
	this->_daysOfWeek = daysOfWeek;
}

void Section::setSemester(Semester* semester){
	this->_dates = semester;
}

void Section::setLocation(Location* building){
	this->_bulding = building;
}

void Section::setCRN(std::string crn){
	this->_crn = crn;
}

/*
 * ======================================================
 * Getters
 * ======================================================
 */


string Section::getSectionName() const{
	return this->_sectionName;
}

string Section::getSectionType() const{
	return this->_sectionType;
}

string Section::getDescription() const{
	return this->_description;
}

Section::Instructor* Section::getInstructor() const{
	return this->_instructor;
}

Section::Week* Section::getWeek() const{
	return this->_daysOfWeek;
}

Section::Semester* Section::getSemester() const{
	return this->_dates;
}

Section::Location* Section::getBuilding() const{
	return this->_bulding;
}

std::string Section::getCRN() const {
	return this->_crn;
}


std::ostream& operator<<(std::ostream& os, const Section& section){
	os << section.getSectionName() << std::endl;
	os << section.getSectionType() << std::endl;
	os << section.getDescription() << std::endl;
	os << section.getCRN() << std::endl;
	
	os << "=== Instructor ==="<< std::endl;
	os << *section.getInstructor() << std::endl;
	
	os << "=== Week ==="<< std::endl;
	os << *section.getWeek() << std::endl;

	os << "=== Semester ==="<< std::endl;
	os << *section.getSemester() << std::endl;

	os << "=== Location ==="<< std::endl;
	os << *section.getBuilding() << std::endl;

	return os;
}

File: Java/illinois/Section.java
package data.scheedule.illinoisold;

import java.util.List;

import org.w3c.dom.Element;

import com.gargoylesoftware.htmlunit.html.HtmlTableCell;
import com.gargoylesoftware.htmlunit.html.HtmlTableRow;

import data.scheedule.base.SectionBase;
import data.scheedule.common.Days;
import data.scheedule.common.Times;
import data.scheedule.common.Days.Day;
import data.scheedule.utils.StringUtils;
import data.scheedule.utils.XmlHelper;

public class Section extends SectionBase {
	
	///////////////////////////////////////////////////////////
	//// Members
	
	private String _sectionName;
	private String _sectionType;
	private Days _days;
	private Times _times;
	private String _startDate;
	private String _endDate;
	private String _description;
	private String _instructor;
	private String _building;
	private String _roomNumber;
	private boolean _isValid;
	
	@Override public String getSectionName() { return _sectionName; }
	@Override public String getSectionType() { return _sectionType; }
	@Override public String getStartDate() { return _startDate; }
	@Override public String getEndDate() { return _endDate; }
	@Override public Days getDays() { return _days; }
	@Override public Times getTimes() { return _times; }
	@Override public String getDescription() { return _description; }
	@Override public String getInstructor() { return _instructor; }
	@Override public String getBuilding() { return _building; }
	@Override public String getRoomNumber() { return _roomNumber; }
	@Override public boolean isValid() { return _isValid; }
	
	private Config getConfig() { return (Config)getBaseConfig(); }
	
	private String _groupByName = null;
	public String getGroupByName() { return _groupByName; }
	public void setGroupByName(String value) { _groupByName = value; }
	
	private String _crn;
	public String getCrn() { return _crn; }
	public void setCrn(String value) { _crn = value; }
	
	public void setSectionName(String value) { _sectionName = value; }
	public void setSectionType(String value) { _sectionType = value; }
			
	///////////////////////////////////////////////////////////
	//// .ctor
	
	public Section(Config config, Element sectionRow) {
		super(config);
		
		_isValid = false;
		
		parseSectionRow(sectionRow);
	}
	
	
	//////////////////////////////////////////////////////////
	//// Helper Functions
	
	private void parseSectionRow(Element sectionRow) {
		
		_crn = XmlHelper.getTextValue(sectionRow, "referenceNumber");
		
		_sectionType = XmlHelper.getTextValue(sectionRow, "sectionType");
		// S# are basically lecture-discussions
		if (_sectionType.startsWith("S"))
		{
			_sectionType = "LCD";
		}
		if (_sectionType.startsWith("C"))
		{
			_sectionType = "CNF";
		}
		_sectionName = XmlHelper.getTextValue(sectionRow, "sectionId");
		_building = XmlHelper.getTextValue(sectionRow, "building");
		_roomNumber = XmlHelper.getTextValue(sectionRow, "roomNumber");
		_instructor = XmlHelper.getTextValue(sectionRow, "instructor");
		_description = XmlHelper.getTextValue(sectionRow, "sectionText");
		_days = parseDays(XmlHelper.getTextValue(sectionRow, "days"));
		_times = parseTimes(XmlHelper.getTextValue(sectionRow, "startTime"), XmlHelper.getTextValue(sectionRow, "endTime"));

		//this._dateRange = new UiucDateRange(XmlHelper.getTextValue(sectionElement, "sectionDateRange"));
		
		_isValid = (_sectionType != null) && (_sectionName != null) && (_days != null)  
						&& (_building != null) && (_roomNumber != null) 
						&& (_instructor != null) && (_description != null);
		
		if (_isValid) {
			System.out.printf("Found Section: %s [%s] from %d-%d on %s in %s room %s taught by %s\n", 
									_sectionName, _sectionType, 
									_times.getStartTime(), _times.getEndTime(),
									_days.toString(),
									_building, _roomNumber, _instructor);
		}
	}

	
	private Times parseTimes(String startTime, String endTime) {
		return new Times(parseTime(startTime), parseTime(endTime));
	}
	
	private int parseTime(String str) {
		int strVal = Integer.parseInt(str.replace(":", "").replace(" AM", "").replace(" PM", ""));
		return strVal + ((strVal < 1200 && str.endsWith("PM")) ? 1200 : 0);
	}
	
	private Days parseDays(String days) {
		Days d = new Days();
		for (int i = 0; i < days.length(); i++) {
			switch (days.charAt(i)) {
			case 'M': d.set(Day.Monday); break;
			case 'T': d.set(Day.Tuesday); break;
			case 'W': d.set(Day.Wednesday); break;
			case 'R': d.set(Day.Thursday); break;
			case 'F': d.set(Day.Friday); break;
			case 'S': d.set(Day.Saturday); break;
			case 'U': d.set(Day.Sunday); break;
			}
		}
		return d;
	}
}
